--- AFLplusplus-2.63c/src/afl-fuzz-queue.c	2020-04-09 16:23:37.000000000 +0800
+++ AFLplusplusSmart/src/afl-fuzz-queue.c	2020-06-28 09:14:34.371236600 +0800
@@ -112,6 +112,10 @@
   q->passed_det = passed_det;
   q->n_fuzz = 1;
   q->trace_mini = NULL;
+  q->validity     = 0;                /* Default validity is 0. */
+  q->parsed       = 0;                /* Structure parsing not attempted. */
+  q->chunk = NULL;
+  q->cached_chunk = NULL;
 
   if (q->depth > afl->max_depth) afl->max_depth = q->depth;
 
@@ -505,10 +509,16 @@
     perf_score = 1;
 
   /* Make sure that we don't go over limit. */
-
   if (perf_score > afl->havoc_max_mult * 100)
     perf_score = afl->havoc_max_mult * 100;
 
+
+  /* Take validity into account */
+  if (afl->smart_mode && ((q->parsed && q->validity >= 50) ||
+                     (!q->parsed && afl->validity_avg >= 50))) {
+    perf_score *= 2;
+  }
+
   return perf_score;
 
 }
